daemon/api.go

```diff
@@ -577,27 +578,53 @@ func waitChange(chg *state.Change) error {
 	return chg.Err()
 }
 
+func ensureUbuntuCore(st *state.State, chg *state.Change, other *state.TaskSet) error {
+	var ss snapstate.SnapState
+	var lastCoreTask *state.Task
+
+	ubuntuCore := "ubuntu-core"
+	err := snapstateGet(st, ubuntuCore, &ss)
+	if err != state.ErrNoState {
+		return nil
+	}
+
+	ts, err := snapstateInstall(st, ubuntuCore, "stable", 0)
+	if err != nil {
+		return err
+	}
+	chg.AddAll(ts)
+
+	tl := ts.Tasks()
+	lastCoreTask = tl[len(tl)-1]
+	other.WaitFor(lastCoreTask)
+	return nil
+}
+
 func (inst *snapInstruction) install() interface{} {
 	flags := snappy.DoInstallGC
 	if inst.LeaveOld {
 		flags = 0
 	}
-	state := inst.overlord.State()
-	state.Lock()
+	st := inst.overlord.State()
+	st.Lock()
+
 	msg := fmt.Sprintf(i18n.G("Install %q snap"), inst.pkg)
 	if inst.Channel != "stable" {
 		msg = fmt.Sprintf(i18n.G("Install %q snap from %q channel"), inst.pkg, inst.Channel)
 	}
-	chg := state.NewChange("install-snap", msg)
-	ts, err := snapstateInstall(state, inst.pkg, inst.Channel, flags)
+	chg := st.NewChange("install-snap", msg)
+	ts, err := snapstateInstall(st, inst.pkg, inst.Channel, flags)
 	if err == nil {
+		if err := ensureUbuntuCore(st, chg, ts); err != nil {
+			return err
```

Comment:
Oops.. return with a lock held. I suggest moving the `snapstateInstall` call above as the first line of the function, and renaming the function to `installSnap`, so we can more easily just return the error here. It will mean we won't need the ts here as well.